File: sdkDemo/sdk/python/mxSdk/packets/base_packet.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
BasePacket - 基础数据包类
描述: 数据包处理的基础类，提供通用的数据包解析功能

作者: RBQ
日期: 2025-06-18
"""

from ..transport.protocol import C, NAK, EOT


class BasePacket:
    def __init__(self):
        self.start = False

    def clear(self) -> None:
        self.start = False

    def is_request_data(self, data: bytes) -> bool:
        if not data or not self.start:
            return False

        if len(data) == 1 and data[0] == C:
            return True

        try:
            if data.decode('utf-8').strip().lower() == 'c':
                return True
        except UnicodeDecodeError:
            pass

        has_c = False
        for byte in data:
            if byte == C:
                has_c = True
            elif byte == NAK or byte == EOT:
                return False
        return has_c

    def is_nak(self, data: bytes) -> bool:
        if not data or not self.start:
            return False

        if len(data) == 1 and data[0] == NAK:
            return True

        try:
            if data.decode('utf-8').strip().lower() == 'r':
                return True
        except UnicodeDecodeError:
            pass

        has_nak = False
        for byte in data:
            if byte == EOT:
                return False
            if byte == NAK:
                has_nak = True
        return has_nak

    def is_eot(self, data: bytes) -> bool:
        if not data or not self.start:
            return False

        if len(data) == 1 and data[0] == EOT:
            return True

        try:
            if data.decode('utf-8').strip().lower() == 'd':
                return True
        except UnicodeDecodeError:
            pass

        return EOT in data

File: mxSdk/Packet/BasePacket.m
//
//  BasePacket.m
//  Inksi
//
//  Created by rbq on 2024/6/17.
//

#import "BasePacket.h"
#import "TransportProtocol.h"

@implementation BasePacket

- (void)setStart:(BOOL)start{
    _start = start;
}

- (void)clear {
    self.start = NO;
}

- (BOOL)isRequestData:(NSData *)data {

    if (!data || !self.start) {
        return NO;
    }

    int length = (int)[data length];
    uint8_t *dataByte = (uint8_t *)[data bytes];

    if (length == 1 && ((dataByte[0] & 0xFF) == C)) {
        return YES;
    }

    NSString *jsonStr = [[[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding]
                               stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
    if([[jsonStr lowercaseString] isEqualToString:@"c"]){
        return YES;
    }
    
    BOOL hasN = NO;
    BOOL hasNAK = NO;
    BOOL hasEOT = NO;
    for (int i = 0; i < length; i++) {
        int value = dataByte[i] & 0xFF;
        if (value == C) {
            hasN = YES;
        } else if (value == NAK) {
            hasNAK = YES;
        } else if (value == EOT) {
            hasEOT = YES;
        }
        // 提前终止循环，如果发现NAK或EOT
        if (hasNAK || hasEOT) {
            return NO;
        }
    }
    return hasN;
}

-(BOOL)isNAK:(NSData *)data {
    
    if (!data || !self.start) {
        return NO;
    }

    int length = (int)[data length];
    uint8_t *dataByte = (uint8_t *)[data bytes];

    if (length == 1 && ((dataByte[0] & 0xFF) == NAK)) {
        return YES;
    }

    NSString *jsonStr = [[[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding]
                               stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
    if([[jsonStr lowercaseString] isEqualToString:@"r"]){
        return YES;
    }
    
    BOOL hasNAK = NO;
    BOOL hasEOT = NO;
    for (int i = 0; i < length; i++) {
        int value = dataByte[i] & 0xFF;
        if (value == NAK) {
            hasNAK = true;
        } else if (value == EOT) {
            hasEOT = true;
            break;  // 提前终止循环，如果发现EOT
        }
    }
    return hasNAK && !hasEOT;
}

-(BOOL)isEOT:(NSData *)data {
    
    if (!data || !self.start) {
        return NO;
    }

    int length = (int)[data length];
    uint8_t *dataByte = (uint8_t *)[data bytes];

    if (length == 1 && ((dataByte[0] & 0xFF) == EOT)) {
        return YES;
    }

    NSString *jsonStr = [[[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding]
                               stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
    if([[jsonStr lowercaseString] isEqualToString:@"d"]){
        return YES;
    }
    
    for (int i = 0; i < length; i++) {
        int value = dataByte[i] & 0xFF;
        if(value == EOT){
            return YES;
        }
    }
    return NO;
}

@end
